src/lib/storage.js

```diff
@@ -149,19 +149,20 @@ class Storage implements IStorageHandler {
       localStream.abort();
       localStream = null; // we force for garbage collector
       self.localStorage.getPackageMetadata(name, (err, info: Package) => {
-        if (_.isNil(err) && info._distfiles && _.isNil(info._distfiles[filename]) === false) {
+        const encodedFilename = encodeURIComponent(filename);
+        if (_.isNil(err) && info._distfiles && _.isNil(info._distfiles[encodedFilename]) === false) {
```

Comment:
Problem number 2: the `info._distFiles` object would have a URI encoded key names, but the `filename` argument was not.

So this condition would fail, and the request for the tarball would 404 (assuming the first problem with 403's was fixed).